### PR TITLE
Sms templates

### PR DESCRIPTION
**Pull request by alxdarksage**
All the SMS messages now have templates, configured in the study object. These are short, but have the same template variables as email messages.

I'm also renaming the "short URLs" so that the variables can be combined in templates (as was recently needed in the "account exists" templates where you can direct the user to reset their password, and/or sign in via email or phone).

### CODE REVIEW

File: app/org/sagebionetworks/bridge/BridgeUtils.java [hunk 1]
```diff
@@ -118,6 +118,7 @@ public static AccountId parseAccountId(String studyId, String identifier) {
                 Iterables.getFirst(commaListToOrderedSet(study.getTechnicalEmail()), ""));
         map.put("consentEmail", 
                 Iterables.getFirst(commaListToOrderedSet(study.getConsentNotificationEmail()), ""));
+        map.put("host", BridgeConfigFactory.getConfig().getHostnameWithPostfix("ws"));
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/BridgeUtils.java:121]: Do we want to add this to BridgeUtilsTest.studyTemplateVariblesWorks?

File: app/org/sagebionetworks/bridge/models/studies/Study.java [hunk 2]
```diff
@@ -267,6 +267,40 @@ static Study create() {
...
+    /**
+     * The template for an SMS message sent to a user that triggers the reset password workflow, 
+     * on an account that only has a phone number.
+     */
+    SmsTemplate getResetPasswordSmsTemplate();
+    void setResetPasswordSmsTemplate(SmsTemplate template);
+    
+    /**
+     * The template for the SMS message sent to a user as part of a sign in via phone workflow.
+     * @return
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/studies/Study.java]: Nitpick: Dangling return annotation.

File: test/org/sagebionetworks/bridge/services/AccountWorkflowServiceTest.java [hunk 3]
```diff
@@ -378,11 +391,12 @@ public void notifyAccountExistsForPhone() throws Exception {
...
                 AccountWorkflowService.EXPIRE_IN_SECONDS);
-        verify(mockNotificationsService).sendSMSMessage(eq(study.getStudyIdentifier()), 
-                eq(TestConstants.PHONE), stringCaptor.capture());
-        String message = stringCaptor.getValue();
+        verify(mockNotificationsService).sendSMSMessage(smsMessageProviderCaptor.capture());
+        
+        String message = smsMessageProviderCaptor.getValue().getSmsRequest().getMessage();
         assertTrue(message.contains("Account for ShortName already exists. Reset password: "));
         assertTrue(message.contains("/rp?study=api&sptoken="+SPTOKEN));
+        assertTrue(message.contains(" or ABC-DEF")); // token is formatted
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/AccountWorkflowServiceTest.java]: Notify account exists for phone also now sends a phone sign-in link. Should we test for that here?

File: test/org/sagebionetworks/bridge/sms/SmsMessageProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        study.setConsentNotificationEmail("consent@email.com,consent2@email.com");
+
+        SmsTemplate template = new SmsTemplate("${studyShortName} ${url} ${supportEmail}");
+        
+        // Create
+        SmsMessageProvider provider = new SmsMessageProvider.Builder()
+            .withStudy(study)
+            .withPhone(TestConstants.PHONE)
+            .withSmsTemplate(template)
+            .withToken("url", "some-url").build();
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/sms/SmsMessageProviderTest.java:37]: Should this include withExpirationPeriod()?

File: test/org/sagebionetworks/bridge/sms/SmsMessageProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+        // Check email
+        PublishRequest request = provider.getSmsRequest();
+        assertEquals("ShortName some-url support@email.com", request.getMessage());
+        
+        assertEquals(study.getShortName(),
+                request.getMessageAttributes().get(BridgeConstants.SENDER_ID).getStringValue());
+        assertEquals(BridgeConstants.SMS_TYPE_TRANSACTIONAL,
+                request.getMessageAttributes().get(BridgeConstants.SMS_TYPE).getStringValue());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/sms/SmsMessageProviderTest.java]: Should we also check the provider's getters as a sanity test?

File: app/org/sagebionetworks/bridge/sms/SmsMessageProvider.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    public Study getStudy() {
+        return study;
+    }
+    public SmsTemplate getTemplate() {
+        return template;
+    }
+    public Phone getPhone() {
+        return phone;
+    }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/sms/SmsMessageProvider.java:41]: Should we have a getter for the token map to facilitate testing?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
Overall looks good. A few notes.

File: app/org/sagebionetworks/bridge/BridgeUtils.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/BridgeUtils.java:121]: Added

File: app/org/sagebionetworks/bridge/models/studies/Study.java (hunk 2 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/studies/Study.java]: removed

File: test/org/sagebionetworks/bridge/services/AccountWorkflowServiceTest.java (hunk 3 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/AccountWorkflowServiceTest.java]: For phone, a token is sent back to be entered into the app. That's what the " or ABC-DEF" test is, reformatted it to " or " + TOKEN

File: test/org/sagebionetworks/bridge/sms/SmsMessageProviderTest.java (hunk 4 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/sms/SmsMessageProviderTest.java:37]: yes, added

File: app/org/sagebionetworks/bridge/sms/SmsMessageProvider.java (hunk 6 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/sms/SmsMessageProvider.java:41]: I didn't use it so I removed it, but I can add it back. It's consistent with MimeTypeEmailProvider. Also added direct tests of the map.

I realized as well that the map should be initialized after you call build(), not after you call the constructor method to get an sms message or email message. So I moved that and fixed the case where null values are added to the map, which will break ImmutableMap.of().

**Review by DwayneJengSage — APPROVED** [review]

**Pull request merged**